pkg/utils/db.go

```diff
@@ -278,17 +279,109 @@ func IsErrBinlogPurged(err error) bool {
 	return ok && e.Code == tmysql.ErrMasterFatalErrorReadingBinlog
 }
 
-// IsErrTableNotExists checks whether err is TableNotExists error
-func IsErrTableNotExists(err error) bool {
-	return IsMySQLError(err, tmysql.ErrNoSuchTable)
+// IsNoSuchThreadError checks whether err is NoSuchThreadError
+func IsNoSuchThreadError(err error) bool {
+	return IsMySQLError(err, tmysql.ErrNoSuchThread)
 }
 
-// IsErrDupEntry checks whether err is DupEntry error
-func IsErrDupEntry(err error) bool {
-	return IsMySQLError(err, tmysql.ErrDupEntry)
+// BaseConn Connect to Database
+type BaseConn struct {
+	DB *sql.DB
 }
 
-// IsNoSuchThreadError checks whether err is NoSuchThreadError
-func IsNoSuchThreadError(err error) bool {
-	return IsMySQLError(err, tmysql.ErrNoSuchThread)
+// SQL is format sql job
+type SQL struct {
+	Query string
+	Args  []interface{}
+}
+
+// Init base BaseConn
+func (conn *BaseConn) Init(dbDSN string) error {
+	db, err := sql.Open("mysql", dbDSN)
+	if err != nil {
+		return terror.DBErrorAdapt(err, terror.ErrDBDriverError, "init")
+	}
+	conn.DB = db
+	return nil
+}
+
+// NormalRetryOperation will retry 100 times
+func (conn *BaseConn) NormalRetryOperation(ctx *tcontext.Context, operateFn func(*tcontext.Context, int) (interface{}, error), retryFn func(int, error) bool) (interface{}, error) {
+	var err error
+	var ret interface{}
+	for i := 0; i < retryFnCount; i++ {
+		ret, err = operateFn(ctx, i)
+		if err != nil {
+			if retryFn(i, err) {
+				continue
+			}
+		}
+		break
+	}
+	return ret, err
+}
+
+// QuerySQL defines query statement, and connect to real DB
+func (conn *BaseConn) QuerySQL(tctx *tcontext.Context, query string) (*sql.Rows, error) {
+	if conn == nil || conn.DB == nil {
+		return nil, terror.ErrDBUnExpect.Generate("database connection not valid")
+	}
+	tctx.L().Debug("Query statement", zap.String("Query", query))
+
+	rows, err := conn.DB.QueryContext(tctx.Context(), query)
+
+	if err != nil {
+		tctx.L().Error("Query statement failed", zap.String("Query", query), log.ShortError(err))
+		return nil, terror.DBErrorAdapt(err, terror.ErrDBQueryFailed, query)
+	}
+	return rows, nil
+}
+
+// ExecuteSQL executes sql on real DB,
+// return
+// 1. failed: (the index of sqls executed error, error)
+// 2. succeed: (len(sqls), nil)
+func (conn *BaseConn) ExecuteSQL(tctx *tcontext.Context, sqls []SQL) (int, error) {
+	if len(sqls) == 0 {
+		return 0, nil
+	}
+	if conn == nil || conn.DB == nil {
+		return 0, terror.ErrDBUnExpect.Generate("database connection not valid")
+	}
+
+	txn, err := conn.DB.Begin()
+	if err != nil {
+		return 0, terror.DBErrorAdapt(err, terror.ErrDBUnExpect)
```

Comment:
Is it needed to include `sqls` in the error here?

BTW, one argument is needed for `ErrDBUnExpect`.